community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.scheduler;
+
+import java.util.Comparator;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.LockSupport;
+
+import org.neo4j.scheduler.JobScheduler.Group;
+import org.neo4j.scheduler.JobScheduler.JobHandle;
+import org.neo4j.time.SystemNanoClock;
+
+final class TimeBasedTaskScheduler implements Runnable
+{
+    private static final long NO_TASKS_PARK = TimeUnit.MINUTES.toNanos( 10 );
+    private static final Comparator<ScheduledJobHandle> DEADLINE_COMPARATOR =
+            Comparator.comparingLong( handle -> handle.nextDeadlineNanos );
+
+    private final SystemNanoClock clock;
+    private final ThreadPoolManager pools;
+    private final PriorityBlockingQueue<ScheduledJobHandle> delayedTasks;
+    private volatile Thread timeKeeper;
+    private volatile boolean stopped;
+
+    TimeBasedTaskScheduler( SystemNanoClock clock, ThreadPoolManager pools )
+    {
+        this.clock = clock;
+        this.pools = pools;
+        delayedTasks = new PriorityBlockingQueue<>( 42, DEADLINE_COMPARATOR );
```

Comment:
While 42 might be the answer to everything, somehow I doubt that includes optimal initial capacity for a blocking priority queue. But I don't what the answer to that would be...

Reply:
The queue grows automatically, and you have to choose something in order to provide a comparator, so here we are.